File: frontend/src/services/ai.ts
import { OHLCV } from '../types/trading';
import { BASE_URL } from './api';

export interface AIStrategy {
  id: string;
  name: string;
  type: 'prediction' | 'classification';
  params: Record<string, any>;
}

export interface PredictionResult {
  predictedPrice: number;
  confidence: number;
  timestamp: string;
}

export const aiService = {
  async trainModel(data: OHLCV[], strategy: AIStrategy) {
    const response = await fetch(`${BASE_URL}/ai/train`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        data,
        strategy
      })
    });
    return response.json();
  },

  async getPrediction(symbol: string, strategy: AIStrategy): Promise<PredictionResult> {
    const response = await fetch(`${BASE_URL}/ai/predict/${symbol}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ strategy })
    });
    return response.json();
  }
}; 

File: backend/app/routers/ai.py
from fastapi import APIRouter, HTTPException
from typing import List, Dict
from app.services.ai_service import AIService

router = APIRouter()
ai_service = AIService()

@router.post("/train")
async def train_model(data: List[Dict], strategy: Dict):
    try:
        result = await ai_service.train_model(data, strategy)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/predict/{symbol}")
async def get_prediction(symbol: str, strategy: Dict):
    try:
        result = await ai_service.get_prediction(symbol, strategy)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e)) 